File: backend/src/services/recommend.ts
import type { Opportunity, User, UserProfile, UserSkill, UserInterest, Skill, Interest } from "@prisma/client";
import { parseJsonArray } from "../utils/deadline";

type ProfileBundle = User & {
  profile: UserProfile | null;
  skills: (UserSkill & { skill: Skill })[];
  interests: (UserInterest & { interest: Interest })[];
};

export type MatchResult = {
  score: number;
  reasons: string[];
};

function overlap(a: string[], b: string[]) {
  const setB = new Set(b.map((x) => x.toLowerCase()));
  return a.filter((x) => setB.has(x.toLowerCase()));
}

function includesLoose(haystack: string | null | undefined, needle: string | null | undefined) {
  if (!haystack || !needle) return false;
  return haystack.toLowerCase().includes(needle.toLowerCase()) || needle.toLowerCase().includes(haystack.toLowerCase());
}

export function scoreOpportunity(user: ProfileBundle, opportunity: Opportunity): MatchResult {
  const reasons: string[] = [];
  let education = 0;
  let field = 0;
  let skills = 0;
  let interest = 0;
  let location = 0;
  let experience = 0;

  const educationLevel = user.profile?.educationLevel;
  if (educationLevel && opportunity.educationLevel) {
    if (includesLoose(opportunity.educationLevel, educationLevel) || opportunity.educationLevel === "any") {
      education = 1;
      reasons.push(`Your education level (${educationLevel}) matches this opportunity.`);
    }
  } else if (!opportunity.educationLevel) {
    education = 0.5;
  }

  const userField = user.profile?.field;
  if (userField && opportunity.field && includesLoose(opportunity.field, userField)) {
    field = 1;
    reasons.push(`Your field of study (${userField}) aligns with ${opportunity.field}.`);
  } else if (userField && includesLoose(opportunity.title + " " + opportunity.description, userField)) {
    field = 0.6;
    reasons.push(`This opportunity is related to your field of study (${userField}).`);
  }

  const userSkills = user.skills.map((s) => s.skill.name);
  const oppSkills = parseJsonArray(opportunity.skills);
  const skillHits = overlap(userSkills, oppSkills.length ? oppSkills : (opportunity.title + " " + opportunity.description).split(/\W+/));
  if (userSkills.length) {
    skills = Math.min(1, skillHits.length / Math.max(2, Math.min(userSkills.length, 4)));
    if (skillHits.length) {
      reasons.push(`Recommended because you have ${skillHits.slice(0, 4).join(", ")} skills.`);
    }
  }

  const userInterests = user.interests.map((i) => i.interest.name);
  const blob = `${opportunity.title} ${opportunity.category} ${opportunity.field ?? ""} ${opportunity.description}`.toLowerCase();
  const interestHits = userInterests.filter((i) => blob.includes(i.toLowerCase()));
  if (userInterests.length) {
    interest = Math.min(1, interestHits.length / Math.max(1, Math.min(userInterests.length, 3)));
    if (interestHits.length) {
      reasons.push(`This matches your interest in ${interestHits.slice(0, 3).join(", ")}.`);
    }
  }

  const preferredLocations = parseJsonArray(user.profile?.preferredLocations);
  if (preferredLocations.length) {
    if (preferredLocations.some((loc) => includesLoose(opportunity.location, loc) || opportunity.location === "remote")) {
      location = 1;
      reasons.push(`The location (${opportunity.location}) matches your preferences.`);
    }
  } else if (user.location && includesLoose(opportunity.location, user.location)) {
    location = 1;
    reasons.push(`This opportunity is in ${opportunity.location}.`);
  }

  const preferredCategories = parseJsonArray(user.profile?.preferredCategories);
  if (preferredCategories.some((c) => c.toLowerCase() === opportunity.category.toLowerCase())) {
    experience = 1;
    reasons.push(`You selected ${opportunity.category} as a preferred category.`);
  } else {
    experience = 0.4;
  }

  const score =
    education * 20 +
    field * 25 +
    skills * 20 +
    interest * 15 +
    location * 10 +
    experience * 10;

  const rounded = Math.round(Math.min(99, Math.max(score, reasons.length ? 42 : 18)));
  if (!reasons.length) {
    reasons.push("Complete your profile to receive more accurate match explanations.");
  }

  return { score: rounded, reasons };
}

export function profileCompleteness(user: ProfileBundle) {
  const checks = [
    Boolean(user.profile?.educationLevel),
    Boolean(user.profile?.field),
    Boolean(user.skills.length),
    Boolean(user.interests.length),
    Boolean(user.profile?.careerGoals),
    Boolean(parseJsonArray(user.profile?.preferredLocations).length),
    Boolean(parseJsonArray(user.profile?.preferredCategories).length),
  ];
  return Math.round((checks.filter(Boolean).length / checks.length) * 100);
}

File: backend/app/services/recommend.py
import re

from sqlalchemy.orm import Session, selectinload

from app.models import Application, Bookmark, Opportunity, OpportunityStatus, User
from app.present import days_remaining
from app.utils import LOCATION_GROUPS, TYPE_LABELS, human_join


def _tokens(value: str) -> list[str]:
    return [token for token in re.split(r"[^a-z0-9+]+", value.lower()) if len(token) > 2]


def _haystack(opp: Opportunity) -> str:
    parts = [
        opp.title,
        opp.short_description,
        opp.description,
        opp.location,
        opp.country,
        " ".join(opp.tags or []),
        " ".join(opp.skills or []),
        opp.organization.name if opp.organization else "",
        opp.category.name if opp.category else "",
    ]
    return " ".join(parts).lower()


def location_match(user: User, opp: Opportunity) -> bool:
    prefs = [item.lower() for item in (user.preferred_locations or [])]
    blob = f"{opp.location} {opp.country}".lower()
    for pref in prefs:
        if pref == "remote" and opp.is_remote:
            return True
        for needle in LOCATION_GROUPS.get(pref, [pref]):
            if needle in blob:
                return True
    return False


def score_opportunity(
    user: User,
    opp: Opportunity,
    bookmark_types: set[str],
    application_types: set[str],
) -> tuple[int, str]:
    score = 0
    text = _haystack(opp)
    preferred = set(user.preferred_categories or [])
    matched_interests: list[str] = []
    matched_skills: list[str] = []

    if opp.opportunity_type in preferred:
        score += 28

    for interest in user.interests or []:
        tokens = _tokens(interest)
        if tokens and any(token in text for token in tokens):
            score += 18
            matched_interests.append(interest)

    opp_skills = {skill.lower() for skill in (opp.skills or []) + (opp.tags or [])}
    for skill in user.skills or []:
        if skill.lower() in opp_skills or any(token in text for token in _tokens(skill)):
            matched_skills.append(skill)
    if matched_skills:
        score += min(36, 12 * len(set(item.lower() for item in matched_skills)))

    if location_match(user, opp):
        score += 14
    if user.education_level and user.education_level in (opp.education_levels or []):
        score += 8
    if opp.opportunity_type in bookmark_types:
        score += 6
    if opp.opportunity_type in application_types:
        score += 4
    if opp.featured:
        score += 2

    bits: list[str] = []
    if matched_interests:
        bits.append(f"you selected {human_join(matched_interests[:3])}")
    if matched_skills:
        bits.append(f"your skills include {human_join(matched_skills[:3])}")
    if opp.opportunity_type in preferred and not matched_interests:
        label = TYPE_LABELS.get(opp.opportunity_type, opp.opportunity_type).lower()
        bits.append(f"you prefer {label}")
    if not bits and location_match(user, opp):
        bits.append("it matches your location preferences")

    if bits:
        reason = "Recommended because " + " and ".join(bits[:2]) + "."
        reason = reason.replace(" and and ", " and ")
    else:
        reason = "Recommended based on your profile."
    return score, reason


def ranked_opportunities(
    db: Session,
    user: User,
    limit: int = 8,
    exclude_ids: set[str] | None = None,
) -> list[tuple[Opportunity, str]]:
    exclude_ids = exclude_ids or set()
    opportunities = (
        db.query(Opportunity)
        .options(selectinload(Opportunity.organization), selectinload(Opportunity.category))
        .filter(Opportunity.status == OpportunityStatus.APPROVED.value)
        .all()
    )
    bookmark_types = {
        row[0]
        for row in db.query(Opportunity.opportunity_type)
        .join(Bookmark, Bookmark.opportunity_id == Opportunity.id)
        .filter(Bookmark.user_id == user.id)
    }
    application_types = {
        row[0]
        for row in db.query(Opportunity.opportunity_type)
        .join(Application, Application.opportunity_id == Opportunity.id)
        .filter(Application.user_id == user.id)
    }
    scored: list[tuple[int, float, Opportunity, str]] = []
    for opp in opportunities:
        if opp.id in exclude_ids:
            continue
        remaining = days_remaining(opp.deadline)
        if remaining is not None and remaining < 0:
            continue
        value, reason = score_opportunity(user, opp, bookmark_types, application_types)
        if value <= 0:
            continue
        created = opp.created_at.timestamp() if opp.created_at else 0
        scored.append((value, created, opp, reason))
    scored.sort(key=lambda item: (item[0], item[1]), reverse=True)
    return [(opp, reason) for _, _, opp, reason in scored[:limit]]
